File: examples/go/report-data/report.go
package main

import (
	"errors"
	"fmt"
	"time"
)

type DataSourceType string
type AnalysisType string
type ReportTemplate string

const (
	DataSourceMySQL      DataSourceType = "MySQL"
	DataSourcePostgreSQL DataSourceType = "PostgreSQL"
	DataSourceCSV        DataSourceType = "CSV"
	DataSourceAPI        DataSourceType = "API"

	AnalysisTimeSeries  AnalysisType = "Time Series Analysis"
	AnalysisCategory    AnalysisType = "Category Analysis"
	AnalysisCorrelation AnalysisType = "Correlation Analysis"
	AnalysisPrediction  AnalysisType = "Prediction Analysis"

	TemplateSummary  ReportTemplate = "Summary Report"
	TemplateDetailed ReportTemplate = "Detailed Report"
	TemplateCustom   ReportTemplate = "Custom Report"
)

type DataSource struct {
	ID       string
	Type     DataSourceType
	Host     string
	Port     string
	Username string
	Password string
	Database string
}

type AnalysisParameters struct {
	StartDate    time.Time
	EndDate      time.Time
	AnalysisType AnalysisType
	Filters      map[string]string
}

type ReportSection struct {
	Name     string
	Included bool
}

type Report struct {
	ID              string
	Title           string
	DataSource      DataSource
	Parameters      AnalysisParameters
	Template        ReportTemplate
	Sections        []ReportSection
	GeneratedAt     time.Time
	Data            []map[string]string
	Summary         string
	Recommendations []string
}

var (
	ErrDataSourceNil        = errors.New("data source cannot be nil")
	ErrHostRequired         = errors.New("host is required")
	ErrPortRequired         = errors.New("port is required")
	ErrUsernameRequired     = errors.New("username is required")
	ErrPasswordRequired     = errors.New("password is required")
	ErrStartDateRequired    = errors.New("start date is required")
	ErrEndDateRequired      = errors.New("end date is required")
	ErrAnalysisTypeRequired = errors.New("analysis type is required")
	ErrTemplateRequired     = errors.New("report template is required")
)

func getDataSources() []DataSource {
	return []DataSource{
		{
			ID:       "ds_001",
			Type:     DataSourceMySQL,
			Host:     "db.example.com",
			Port:     "3306",
			Username: "analyst",
			Database: "sales_data",
		},
		{
			ID:       "ds_002",
			Type:     DataSourcePostgreSQL,
			Host:     "analytics.example.com",
			Port:     "5432",
			Username: "analyst",
			Database: "customer_data",
		},
		{
			ID:       "ds_003",
			Type:     DataSourceCSV,
			Host:     "files.example.com",
			Port:     "",
			Username: "analyst",
			Database: "exports",
		},
		{
			ID:       "ds_004",
			Type:     DataSourceAPI,
			Host:     "api.example.com",
			Port:     "443",
			Username: "analyst",
			Database: "",
		},
	}
}

func getReportTemplates() []ReportTemplate {
	return []ReportTemplate{
		TemplateSummary,
		TemplateDetailed,
		TemplateCustom,
	}
}

func getReportSections() []ReportSection {
	return []ReportSection{
		{Name: "Overview", Included: true},
		{Name: "Detailed Data", Included: true},
		{Name: "Graphs", Included: true},
		{Name: "Recommendations", Included: true},
	}
}

func generateReport(dataSource DataSource, params AnalysisParameters, template ReportTemplate, sections []ReportSection) (*Report, error) {
	if err := validateDataSource(&dataSource); err != nil {
		return nil, err
	}

	if err := validateParameters(&params); err != nil {
		return nil, err
	}

	if template == "" {
		return nil, ErrTemplateRequired
	}

	// Simulate data retrieval and analysis
	report := &Report{
		ID:              fmt.Sprintf("report_%d", time.Now().UnixNano()),
		Title:           fmt.Sprintf("%s Analysis Report", params.AnalysisType),
		DataSource:      dataSource,
		Parameters:      params,
		Template:        template,
		Sections:        sections,
		GeneratedAt:     time.Now().UTC(),
		Data:            generateSampleData(params),
		Summary:         generateSummary(params),
		Recommendations: generateRecommendations(params),
	}

	return report, nil
}

func validateDataSource(ds *DataSource) error {
	if ds == nil {
		return ErrDataSourceNil
	}
	if ds.Host == "" {
		return ErrHostRequired
	}
	if ds.Port == "" && ds.Type != DataSourceCSV {
		return ErrPortRequired
	}
	if ds.Username == "" {
		return ErrUsernameRequired
	}
	if ds.Password == "" {
		return ErrPasswordRequired
	}
	return nil
}

func validateParameters(params *AnalysisParameters) error {
	if params.StartDate.IsZero() {
		return ErrStartDateRequired
	}
	if params.EndDate.IsZero() {
		return ErrEndDateRequired
	}
	if params.AnalysisType == "" {
		return ErrAnalysisTypeRequired
	}
	return nil
}

func generateSampleData(params AnalysisParameters) []map[string]string {
	// Generate sample data based on analysis type
	var data []map[string]string

	switch params.AnalysisType {
	case AnalysisTimeSeries:
		// Generate time series data
		currentDate := params.StartDate
		for currentDate.Before(params.EndDate) || currentDate.Equal(params.EndDate) {
			value := 100 + (currentDate.Day() % 20) // Simple variation
			change := "+5%"
			if currentDate.Day()%3 == 0 {
				change = "-2%"
			}

			data = append(data, map[string]string{
				"Date":   currentDate.Format("2006-01-02"),
				"Value":  fmt.Sprintf("%d", value),
				"Change": change,
			})

			currentDate = currentDate.AddDate(0, 0, 1)
		}

	case AnalysisCategory:
		// Generate category data
		categories := []string{"Electronics", "Clothing", "Food", "Books", "Home"}
		for _, category := range categories {
			value := 100 + (len(category) * 10)
			percentage := float64(value) / 500.0 * 100.0

			data = append(data, map[string]string{
				"Category":   category,
				"Value":      fmt.Sprintf("%d", value),
				"Percentage": fmt.Sprintf("%.1f%%", percentage),
			})
		}

	case AnalysisCorrelation:
		// Generate correlation data
		for i := 0; i < 10; i++ {
			x := 10 + i*5
			y := 20 + i*3 + (i%3)*2

			data = append(data, map[string]string{
				"X": fmt.Sprintf("%d", x),
				"Y": fmt.Sprintf("%d", y),
			})
		}

	case AnalysisPrediction:
		// Generate prediction data
		currentDate := params.StartDate
		for i := 0; i < 10; i++ {
			actual := 100 + i*5
			predicted := actual + (i%3)*2 - 1

			data = append(data, map[string]string{
				"Date":       currentDate.Format("2006-01-02"),
				"Actual":     fmt.Sprintf("%d", actual),
				"Predicted":  fmt.Sprintf("%d", predicted),
				"Difference": fmt.Sprintf("%d", predicted-actual),
			})

			currentDate = currentDate.AddDate(0, 0, 1)
		}
	}

	return data
}

func generateSummary(params AnalysisParameters) string {
	switch params.AnalysisType {
	case AnalysisTimeSeries:
		return "This report analyzes the time series data for the selected period. The data shows a general upward trend with some fluctuations. Key insights include a 15% overall growth and three significant spikes in activity."
	case AnalysisCategory:
		return "This category analysis reveals that Electronics and Clothing are the top performing categories, accounting for 45% of total value. Food and Books show steady performance, while Home category has shown recent decline."
	case AnalysisCorrelation:
		return "The correlation analysis indicates a strong positive relationship between variables X and Y (r=0.85). This suggests that changes in X are likely to cause proportional changes in Y."
	case AnalysisPrediction:
		return "The prediction model shows good accuracy with an average error margin of 3.2%. The model predicts continued growth in the next quarter, with a 12% increase expected."
	default:
		return "This report provides an analysis of the selected data based on the specified parameters."
	}
}

func generateRecommendations(params AnalysisParameters) []string {
	switch params.AnalysisType {
	case AnalysisTimeSeries:
		return []string{
			"Invest in marketing during the identified peak periods",
			"Consider seasonal inventory adjustments based on the pattern",
			"Implement targeted promotions during low activity periods",
		}
	case AnalysisCategory:
		return []string{
			"Increase investment in the Electronics category",
			"Review pricing strategy for the Home category",
			"Explore cross-promotion opportunities between top categories",
		}
	case AnalysisCorrelation:
		return []string{
			"Leverage the strong correlation to optimize resource allocation",
			"Consider A/B testing to validate the relationship",
			"Monitor for any changes in the correlation strength",
		}
	case AnalysisPrediction:
		return []string{
			"Prepare for the predicted growth in the next quarter",
			"Review current capacity to ensure it can handle the expected increase",
			"Consider early procurement to avoid potential supply constraints",
		}
	default:
		return []string{
			"Review the data for any unexpected patterns",
			"Consider additional analysis to gain deeper insights",
			"Share findings with relevant stakeholders",
		}
	}
}

File: examples/node/report-data/src/report.ts
type DataSourceType = string;
type AnalysisType = string;
type ReportTemplate = string;

export type DataSource = {
  id: string;
  type: DataSourceType;
  host: string;
  port: string;
  username: string;
  password: string;
  database: string;
};

export type AnalysisParameters = {
  startDate: Date | null;
  endDate: Date | null;
  analysisType: AnalysisType;
  filters: Record<string, string>;
};

export type ReportSection = {
  name: string;
  included: boolean;
};

const dataSourceMySQL: DataSourceType = 'MySQL' as const;
const dataSourcePostgreSQL: DataSourceType = 'PostgreSQL' as const;
const dataSourceCSV: DataSourceType = 'CSV' as const;
const dataSourceAPI: DataSourceType = 'API' as const;

export const analysisTimeSeries: AnalysisType = 'Time Series Analysis' as const;
export const analysisCategory: AnalysisType = 'Category Analysis' as const;
export const analysisCorrelation: AnalysisType =
  'Correlation Analysis' as const;
export const analysisPrediction: AnalysisType = 'Prediction Analysis' as const;

const templateSummary: ReportTemplate = 'Summary Report' as const;
const templateDetailed: ReportTemplate = 'Detailed Report' as const;
const templateCustom: ReportTemplate = 'Custom Report' as const;

type ReportType = {
  id: string;
  title: string;
  dataSource: DataSource;
  parameters: AnalysisParameters;
  template: ReportTemplate;
  sections: ReportSection[];
  generatedAt: Date;
  data: Record<string, string>[];
  summary: string;
  recommendations: string[];
};

const validateDataSource = (ds: DataSource): null => {
  if (ds.host === '') {
    throw new Error('Host is required');
  }
  if (ds.port === '' && ds.type !== dataSourceCSV) {
    throw new Error('Port is required');
  }
  if (ds.username === '') {
    throw new Error('Username is required');
  }
  if (ds.password === '') {
    throw new Error('Password is required');
  }
  return null;
};

const validateParameters = (params: AnalysisParameters): null => {
  if (!params.startDate || params.startDate.getTime() === 0) {
    throw new Error('Start date is required');
  }
  if (!params.endDate || params.endDate.getTime() === 0) {
    throw new Error('End date is required');
  }
  if (params.analysisType === '') {
    throw new Error('Analysis type is required');
  }
  return null;
};

const generateSampleData = (
  params: AnalysisParameters,
): Record<string, string>[] => {
  // Generate sample data based on analysis type
  const data: Record<string, string>[] = [];

  switch (params.analysisType) {
    case analysisTimeSeries: {
      // Generate time series data
      let currentDate = params.startDate;
      while (currentDate && params.endDate && currentDate < params.endDate) {
        const value = 100 + (currentDate.getDate() % 20); // Simple variation
        let change = '+5%';
        if (currentDate.getDate() % 3 === 0) {
          change = '-2%';
        }

        data.push({
          Date: currentDate.toISOString().split('T')[0],
          Value: value.toString(),
          Change: change,
        });

        currentDate = new Date(currentDate.getTime() + 24 * 60 * 60 * 1000);
      }
      break;
    }

    case analysisCategory: {
      // Generate category data
      const categories = ['Electronics', 'Clothing', 'Food', 'Books', 'Home'];
      for (const category of categories) {
        const value = 100 + category.length * 10;
        const percentage = (value / 500.0) * 100.0;

        data.push({
          Category: category,
          Value: value.toString(),
          Percentage: percentage.toString(),
        });
      }
      break;
    }
    case analysisCorrelation: {
      // Generate correlation data
      for (let i = 0; i < 10; i++) {
        const x = 10 + i * 5;
        const y = 20 + i * 3 + (i % 3) * 2;

        data.push({
          X: x.toString(),
          Y: y.toString(),
        });
      }
      break;
    }

    case analysisPrediction: {
      // Generate prediction data
      let currentDate = params.startDate;
      if (currentDate) {
        for (let i = 0; i < 10; i++) {
          const actual = 100 + i * 5;
          const predicted = actual + (i % 3) * 2 - 1;

          data.push({
            Date: currentDate.toISOString().split('T')[0],
            Actual: actual.toString(),
            Predicted: predicted.toString(),
            Difference: (predicted - actual).toString(),
          });

          currentDate = new Date(currentDate.getTime() + 24 * 60 * 60 * 1000);
        }
      }
      break;
    }
  }

  return data;
};

const generateSummary = (params: AnalysisParameters): string => {
  switch (params.analysisType) {
    case analysisTimeSeries:
      return 'This report analyzes the time series data for the selected period. The data shows a general upward trend with some fluctuations. Key insights include a 15% overall growth and three significant spikes in activity.';
    case analysisCategory:
      return 'This category analysis reveals that Electronics and Clothing are the top performing categories, accounting for 45% of total value. Food and Books show steady performance, while Home category has shown recent decline.';
    case analysisCorrelation:
      return 'The correlation analysis indicates a strong positive relationship between variables X and Y (r=0.85). This suggests that changes in X are likely to cause proportional changes in Y.';
    case analysisPrediction:
      return 'The prediction model shows good accuracy with an average error margin of 3.2%. The model predicts continued growth in the next quarter, with a 12% increase expected.';
    default:
      return 'This report provides an analysis of the selected data based on the specified parameters.';
  }
};

const generateRecommendations = (params: AnalysisParameters): string[] => {
  switch (params.analysisType) {
    case analysisTimeSeries:
      return [
        'Invest in marketing during the identified peak periods',
        'Consider seasonal inventory adjustments based on the pattern',
        'Implement targeted promotions during low activity periods',
      ];
    case analysisCategory:
      return [
        'Increase investment in the Electronics category',
        'Review pricing strategy for the Home category',
        'Explore cross-promotion opportunities between top categories',
      ];
    case analysisCorrelation:
      return [
        'Leverage the strong correlation to optimize resource allocation',
        'Consider A/B testing to validate the relationship',
        'Monitor for any changes in the correlation strength',
      ];
    case analysisPrediction:
      return [
        'Prepare for the predicted growth in the next quarter',
        'Review current capacity to ensure it can handle the expected increase',
        'Consider early procurement to avoid potential supply constraints',
      ];
    default:
      return [
        'Review the data for any unexpected patterns',
        'Consider additional analysis to gain deeper insights',
        'Share findings with relevant stakeholders',
      ];
  }
};

export const getDataSources = (): Partial<DataSource>[] => {
  return [
    {
      id: 'ds_001',
      type: dataSourceMySQL,
      host: 'db.example.com',
      port: '3306',
      username: 'analyst',
      database: 'sales_data',
    },
    {
      id: 'ds_002',
      type: dataSourcePostgreSQL,
      host: 'analytics.example.com',
      port: '5432',
      username: 'analyst',
      database: 'customer_data',
    },
    {
      id: 'ds_003',
      type: dataSourceCSV,
      host: 'files.example.com',
      port: '',
      username: 'analyst',
      database: 'exports',
    },
    {
      id: 'ds_004',
      type: dataSourceAPI,
      host: 'api.example.com',
      port: '443',
      username: 'analyst',
      database: '',
    },
  ];
};

export const generateReport = (
  dataSource: DataSource,
  params: AnalysisParameters,
  template: ReportTemplate,
  sections: ReportSection[],
): ReportType | null => {
  if (
    validateDataSource(dataSource) !== null ||
    validateParameters(params) !== null
  ) {
    return null;
  }

  if (template === '') {
    throw new Error('Template is required');
  }

  // Simulate data retrieval and analysis
  const report: ReportType = {
    id: `report_${Date.now()}`,
    title: `${params.analysisType} Analysis Report`,
    dataSource: dataSource,
    parameters: params,
    template: template,
    sections: sections,
    generatedAt: new Date(),
    data: generateSampleData(params),
    summary: generateSummary(params),
    recommendations: generateRecommendations(params),
  };

  return report;
};

export const getReportTemplates = (): ReportTemplate[] => {
  return [templateSummary, templateDetailed, templateCustom];
};

export const getReportSections = (): ReportSection[] => {
  return [
    { name: 'Overview', included: true },
    { name: 'Detailed Data', included: true },
    { name: 'Graphs', included: true },
    { name: 'Recommendations', included: true },
  ];
};
